src/xapiwrapper.js

```diff
@@ -1369,6 +1367,186 @@ function isDate(date) {
+            this.async = true;
+            this.headers = {};
+        };
+        XMLHttpRequest.prototype = {
+
+            open: function(method, url, async)
+            {
+
+                if (async === false) {
+                    throw "ADL xAPIWrapper does not support synchronous http requests in node";
+                }
+
+                this.method = method;
+                this.url = url;
+                this.withCredentials = true;
+                this.crossDomain = true;
+                this.responseText = "";
+                this.responseJSON = null;
+                this.readyState = 0;
+                this.status = 0;
+                this.onreadystatechange = function() {};
+                this.onerror = function(error) {};
+                this.onload = function() {};
+            },
+
+            setRequestHeader: function(name, value)
+            {
+                this.headers[name] = value;
+            },
+
+            send: function(data)
+            {
+                var http = this.url.includes("https:") ? require('https') : require("http");
+                var options = {
+                    method: this.method,
+                    headers: this.headers
+                };
+                var parsedUrl = parseUrl(this.url);
+                for (var k in parsedUrl) {
+                    options[k] = parsedUrl[k];
+                }
+                var req = http.request(options, function (res) {
+                    res.setEncoding('utf8');
+                    this.status = res.statusCode;
+                    res.on('data', function (d) {
+                        this.responseText+=d;
+                    }.bind(this));
+                    res.on('end', function () {
+                        this.readyState = 4;
+                        try {
+                            this.responseJSON = JSON.parse(this.responseText);
+                        } catch(error) {
+                            this.responseJSON = null;
+                        }
+                        this.onload();
+                    }.bind(this));
+                }.bind(this));
+                req.on('error', function (e) {
+                    this.readyState = 4;
+                    this.onerror();
+                }.bind(this));
+                req.end(data);
+            }
+
+        };
+
+     })();
+
+    /**
+     * Cross environment http requests
+     * @param  {string} method   the http request method (ex: "PUT", "GET")
+     * @param  {string} url   the url to the request (ex: ADL.XAPIWrapper.lrs.endpoint + "statements")
+     * @param  {object} [headers]   headers to include in the request
+     * @param  {string} [data]   the body of the request, if there is one
+     * @param  {object} [options] request options
+     * @param  {function} [complete ] completion callback function
+     * @return {object} xhr response object
+     */
+    function http(method, url, headers, data, options, complete)
```

Comment:
I'm not entirely sure about the definition of the `http` function parameters, but the intent of the code (as a method of performing cross domain http request requests in different environments) seems easier to read when isolated like this. The `http` function should hopefully exhibit identical behaviour to the original code which was nested inside the `XHR_request` function, as it's almost a verbatim copy. I have left the LRS specific http request modifications in the `XHR_request` function to create a distinct line between LRS stuff and the http request stuff. I hope that makes sense? This bit will need the most testing.